### PR TITLE
Fix fetch tags

### PR DESCRIPTION
**Pull request by zacharya19**
Quick fix for #371.
Not sure it's the best way, but this bug is very disturbing to me, so I fixed it quickly.

In tags we do not need to check hash, we need to check if it's name exists in the local storer, so I saved all local tags in slice and then if the reference is a tag, I say it's exists if it's in the slice.

Better idea?

### CODE REVIEW

File: remote.go [hunk 1]
```diff
@@ -329,6 +329,21 @@ func getWants(
 		}
 	}
 
+	tags := []plumbing.Hash{}
```
Comment (smola) [inline remote.go]: Prefer `var tags []plumbing.Hash` initialization.

File: remote.go [hunk 2]
```diff
@@ -329,6 +329,21 @@ func getWants(
...
+	localIter, err := localStorer.IterReferences()
+	if err != nil {
+		return nil, err
+	}
+
+	localIter.ForEach(func(ref *plumbing.Reference) error {
```
Comment (smola) [inline remote.go]: Check return error.

**Comment by smola** [comment]
@zacharya19 Thanks for the contribution!

I wonder how this works. If I understood correctly, this implies that the repository contains a tag returned by `IterReferences()` that points to a hash which does not exist in the `localStorer`.

That should not happen at all. Could you give more details about a repository exhibiting this error? Did you clone it with go-git from the beginning? Or did you clone with git and then pulled with go-git? I suspect that if this fix works, it's only because there is a related bug that has lead the repository into an inconsistent state.

**Comment by zacharya19** [comment]
@smola I use PlainOpen to open a repo I cloned with git cli.
But there is something strange here, with my fix I get new tags with the same info but other name (I have tag named "release-1.23.1" and after I run the code, my local git has "release-1.23.1" and "1.23.1" with the same hash and message.
I'm not really sure why.

**Comment by zacharya19** [comment]
BTW - the tag could not by exists locally if the hash does not exists locally, am I wrong?

**Comment by smola** [comment]
@zacharya19 This code implies that tags have hashes of objects that are not in the object storage. A tag is either a tag object (in the object storage) or a reference to a non-tag object (in the object storage too). So if there are cases where `IterReferences()` return tags pointing to hashes that are not in the object storage (`objectExists(localStorer, hash) == false`) it might indicate a bug previos in the execution.

```
if !ref.IsTag() {
	exists, err = objectExists(localStorer, hash)
	if err != nil {
		return err
	}
} else {
	exists = hashInList(hash, tags)
}
```

**Comment by zacharya19** [comment]
Not sure I got it, since the problem here is the opposite, on the remote we see tags that point to an hash that exists locally, but the tag it self does not exist.

**Comment by smola** [comment]
@zacharya19 Ok. This can happen with lightweight tags. Although, this still wouldn't fix the case where a tag `T1`already exist locally pointing to hash `H1` and the server contains a new hash `T2` pointing to `H1` too.

**Comment by zacharya19** [comment]
@smola I made some changes, let me know what you think.

**Comment by mcuadros** [comment]
@smola

**Comment by smola** [comment]
We should add a test for this though.

**Comment by smola** [comment]
I'm working on a proper test before moving on.

**Comment by zacharya19** [comment]
@smola I added basic test, let me know what you think.

**Comment by mcuadros** [comment]
@zacharya19 thanks for your efforts and your time, but I created another PR https://github.com/src-d/go-git/pull/485 with a canonical solution.

**Pull request closed without merging**